### PR TITLE
Add scroll area for share links.

### PR DESCRIPTION
**Pull request by camilasan**
share-dialog

### CODE REVIEW

**Comment by camilasan** [comment]
@jancborchardt what do you think about these changes?

**Review by jancborchardt — CHANGES_REQUESTED** [review]
> version 2:
> put it all in one scroll?

I'd say this should definitely be in one scroll area from the first version, 2 different scroll areas is always very strange. Otherwise good!

**Comment by camilasan** [comment]
> I'd say this should definitely be in one scroll area from the first version, 2 different scroll areas is always very strange. Otherwise good!

Ok, what is the logic when you are add a new link or user? they are just added regardless to the bottom of the list mixed? Or links are kept on top?

**Comment by camilasan** [comment]
> > I'd say this should definitely be in one scroll area from the first version, 2 different scroll areas is always very strange. Otherwise good!
> 
> Ok, what is the logic when you are add a new link or user? they are just added regardless to the bottom of the list mixed? Or links are kept on top?

this is actually fine.

**Comment by tobiasKaminsky** [comment]
> Ok, what is the logic when you are add a new link or user? they are just added regardless to the bottom of the list mixed? Or links are kept on top?

On Android we have this sorting:
- first public/email shares
- then all user/federated/talk/conversation/etc shares
- at bottom internal share

All shares are sorted by share date from new to old.

**Comment by camilasan** [comment]
> On Android we have this sorting:
> 
>     * first public/email shares
> 
>     * then all user/federated/talk/conversation/etc shares
> 
>     * at bottom internal share
> 
> 
> All shares are sorted by share date from new to old.

thank you. I am hoping we can do this another time.

File: src/gui/sharedialog.cpp [hunk 1]
```diff
@@ -171,33 +177,35 @@ ShareLinkWidget *ShareDialog::addLinkShareWidget(const QSharedPointer<LinkShare>
...
+        _emptyShareLinkWidget = new ShareLinkWidget(_accountState->account(), _sharePath, _localPath, _maxSharingPermissions, _ui->scrollArea);
         _linkWidgetList.append(_emptyShareLinkWidget);
-
-        connect(_emptyShareLinkWidget, &ShareLinkWidget::resizeRequested, this, &ShareDialog::slotAdjustScrollWidgetSize);
         connect(this, &ShareDialog::toggleShareLinkAnimation, _emptyShareLinkWidget, &ShareLinkWidget::slotToggleShareLinkAnimation);
         connect(_emptyShareLinkWidget, &ShareLinkWidget::createLinkShare, this, &ShareDialog::slotCreateLinkShare);
 
         connect(_emptyShareLinkWidget, &ShareLinkWidget::createPassword, this, &ShareDialog::slotCreatePasswordForLinkShare);
-
+
```
Comment (mgallien) [inline src/gui/sharedialog.cpp:204]: you have extra spaces

File: src/gui/sharedialog.cpp [hunk 2]
```diff
@@ -171,33 +177,35 @@ ShareLinkWidget *ShareDialog::addLinkShareWidget(const QSharedPointer<LinkShare>
...
 
     _ui->verticalLayout->insertWidget(_linkWidgetList.size() + 1, linkShareWidget);
+    _scrollAreaLayout->addWidget(linkShareWidget);
+    
+    // TO DO - the count is right but the bkg does not change
+    //linkShareWidget->setBackgroundRole(_scrollAreaLinksLayout->count() % 2 == 0 ? QPalette::Base : QPalette::AlternateBase);
+    
     linkShareWidget->setupUiOptions();
-
+
```
Comment (mgallien) [inline src/gui/sharedialog.cpp:191]: you have extra spaces

File: src/gui/sharedialog.cpp [hunk 3]
```diff
@@ -303,21 +314,17 @@ void ShareDialog::showSharingUi()
...
+        
         // Connect styleChanged events to our widget, so it can adapt (Dark-/Light-Mode switching)
         connect(this, &ShareDialog::styleChanged, _userGroupWidget, &ShareUserGroupWidget::slotStyleChanged);
 
         _ui->verticalLayout->insertWidget(1, _userGroupWidget);
-        _userGroupWidget->getShares();
+        _scrollAreaLayout->addLayout(_userGroupWidget->shareUserGroupLayout());
     }
-
+
```
Comment (mgallien) [inline src/gui/sharedialog.cpp:327]: you have extra spaces

File: src/gui/sharedialog.cpp [hunk 4]
```diff
@@ -230,19 +240,20 @@ void ShareDialog::slotSharesFetched(const QList<QSharedPointer<Share>> &shares)
         QSharedPointer<LinkShare> linkShare = qSharedPointerDynamicCast<LinkShare>(share);
         addLinkShareWidget(linkShare);
     }
-
+
```
Comment (mgallien) [inline src/gui/sharedialog.cpp:243]: you have extra spaces

File: src/gui/shareusergroupwidget.cpp [hunk 5]
```diff
@@ -247,17 +249,17 @@ void ShareUserGroupWidget::slotShareCreated(const QSharedPointer<Share> &share)
...
-    QScrollArea *scrollArea = _parentScrollArea;
-
-    auto newViewPort = new QWidget(scrollArea);
-    auto layout = new QVBoxLayout(newViewPort);
-    layout->setContentsMargins(0, 0, 0, 0);
     int x = 0;
-    int height = 0;
     QList<QString> linkOwners({});
-
+
```
Comment (mgallien) [inline src/gui/shareusergroupwidget.cpp:254]: you have extra spaces

**Review by mgallien — COMMENTED** [review]
I am going now for a second review round

**Comment by mgallien** [comment]
there are still issues with the layout
see screenshot 
image
the button on the share link is not entirely visible

**Comment by mgallien** [comment]
when you share it with an user the first time the window is quite small
can you make it taller ?

File: src/gui/sharedialog.cpp [hunk 6]
```diff
@@ -150,13 +150,19 @@ ShareDialog::ShareDialog(QPointer<AccountState> accountState,
...
 
 ShareLinkWidget *ShareDialog::addLinkShareWidget(const QSharedPointer<LinkShare> &linkShare)
 {
-    _linkWidgetList.append(new ShareLinkWidget(_accountState->account(), _sharePath, _localPath, _maxSharingPermissions, this));
-
-    const auto linkShareWidget = _linkWidgetList.at(_linkWidgetList.size() - 1);
+    _linkWidgetList.append(new ShareLinkWidget(_accountState->account(), _sharePath, _localPath, _maxSharingPermissions, _ui->scrollArea));
+    
+    const auto index = _linkWidgetList.size() - 1;
+    const auto linkShareWidget = _linkWidgetList.at(index);
```
Comment (allexzander) [inline src/gui/sharedialog.cpp:165]: @camilasan Do you need that `index` variable? Also, won't this work similarly `const auto linkShareWidget = _linkWidgetList.last();` ?

File: src/gui/sharedialog.cpp [hunk 7]
```diff
@@ -230,19 +240,20 @@ void ShareDialog::slotSharesFetched(const QList<QSharedPointer<Share>> &shares)
...
 void ShareDialog::slotAdjustScrollWidgetSize()
 {
-    int count = this->findChildren<ShareLinkWidget *>().count();
-    _ui->scrollArea->setVisible(count > 0);
-    if (count > 0 && count <= 3) {
-        _ui->scrollArea->setFixedHeight(_ui->scrollArea->widget()->sizeHint().height());
-    }
-    _ui->scrollArea->setFrameShape(count > 3 ? QFrame::StyledPanel : QFrame::NoFrame);
+    const auto count = _scrollAreaLayout->count();
+    const auto height = _linkWidgetList.size() > 0 ? _linkWidgetList.at(_linkWidgetList.size() - 1)->sizeHint().height() : 0;
```
Comment (allexzander) [inline src/gui/sharedialog.cpp:252]: @camilasan `_linkWidgetList.at(_linkWidgetList.size() - 1)` -> `_linkWidgetList.last()`

File: src/gui/shareusergroupwidget.cpp [hunk 8]
```diff
@@ -247,17 +249,17 @@ void ShareUserGroupWidget::slotShareCreated(const QSharedPointer<Share> &share)
...
     int x = 0;
-    int height = 0;
     QList<QString> linkOwners({});
-
+    
     ShareUserLine *justCreatedShareThatNeedsPassword = nullptr;
-
+       
+    QLayoutItem *shareUserLine;
+    while ((shareUserLine = _shareUserGroup->takeAt(0)) != nullptr) {
```
Comment (allexzander) [inline src/gui/shareusergroupwidget.cpp:258]: @camilasan You may make it shorter and fix clang-tidy complaints:

`while ((shareUserLine = _shareUserGroup->takeAt(0)) != nullptr)` -> `while (QLayoutItem *shareUserLine = _shareUserGroup->takeAt(0))`

File: src/gui/sharedialog.cpp (hunk 6 above)
Comment (camilasan) [reply src/gui/sharedialog.cpp:165]: leftover from something I was trying before but now I rewrote it better (imho) and I don't need last() or at().

**Comment by tobiasKaminsky** [comment]
> thank you. I am hoping we can do this another time.

sure, just create an issue, so we can track it 👍

**Comment by tobiasKaminsky** [comment]
image

I cannot see the "share link" at all, with appImage from above.

**Comment by camilasan** [comment]
> I cannot see the "share link" at all, with appImage from above.

I don't get it. This is what I see:
sharedialog

**Review by allexzander — APPROVED** [review]

**Comment by tobiasKaminsky** [comment]
If this is working for you and Alex, it might be only a problem on my side, then please ignore me ;)

**Comment by claucambra** [comment]
The share button is working on my side FWIW

**Review by claucambra — APPROVED** [review]

**Comment by mgallien** [comment]
@jancborchardt do you still want more changes or are you OK with the solution done by @camilasan ?

**Review by mgallien — APPROVED** [review]
thanks
works for me

**Comment by mgallien** [comment]
@camilasan can you rebase on your clone because there is a conflict ?

**Pull request merged**